Bugfix/fix arp response is not received

morantr opened the pull request:
It is observed that sometimes mon_rssi_want to send arp request, but it doesn't have the
ipv4 of the client. If we don't have the station ipv4, the arp request will not get a response, and sending it to wildcard IP doesn't work apparently.

This PR adds the following:
Fix bug in the updating of the station IP on the Agent.
Fix ipv4 that was unset on start monitoring message.

arnout commented on common/beerocks/bcl/include/bcl/network/network_utils.h:
```diff
@@ -101,8 +101,8 @@ class network_utils {
 
-    //temp
     static std::string get_mac_from_arp_table(const std::string &ipv4);
+    static std::string get_ipv4_from_arp_table(const std::string &mac);
```
Please, no new functions which take a mac-as-string as argument...

arnout reviewed (approved):
I'll approve it, but I wonder if this really works in practice? How is the MAC address going to end up in the ARP table - that only happens if either we sent an IP packet to the STA, or the STA sent an IP packet to us. That is only going to happen in practice if we're the DHCP server - but normally, only the gateway runs a DHCP server.

That said, if it works, I'll accept it.

vitalybu commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -419,23 +419,24 @@ static int parseRoutes(struct nlmsghdr *nlHdr, std::shared_ptr<route_info> rtInf
+    // IP address       HW type     Flags       HW address            Mask     Device
+    // 192.168.12.31    0x1         0x2         00:09:6b:00:02:03     *        eth0
+    // 192.168.12.70    0x1         0x2         00:01:02:38:4c:85     *        eth0
     std::ifstream arp_table("/proc/net/arp");
```
This is definitely not the way to do this.
If we decide to go this way, should use netlink sockets instead.

kantera800 commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -1030,21 +1030,16 @@ bool network_utils::arp_send(const std::string &iface, const std::string &dst_ip
-
-    // If the destination IP is empty, there is no point sending the arp, therefore replace the
-    // with broadcast IP, so all clients will receive it and answer, but since the request is
-    // being sent to a specific mac address, then only the requested client will answer.
```
You need to have a better reason than "not working well"  to remove this code.
It used to work so clearly something broke , or maybe it does work on some clients (another reason to keep it) 
Also , this commit is not related to this PR and you cannot close  #1056 until this issue is resolved or at least debugged.

kantera800 commented on agent/src/beerocks/monitor/monitor_rssi.cpp:
```diff
@@ -337,6 +337,12 @@ void monitor_rssi::process()
+                    sta_node->get_ipv4() == network_utils::ZERO_IP_STRING) {
+                    auto ipv4 = network_utils::get_ipv4_from_arp_table(arp_dst_mac);
+                    sta_node->set_ipv4(ipv4);
+                }
```
instead of adding this w/a  you should  fix the real problem.
i see that in the past you already added a w/a for this issue https://github.com/prplfoundation/prplMesh/commit/873bf9ed6a3fba6322e8eb001071e991bf108bf1 , but then the `Wait until ipv4 will be received on DHCP event` part was removed .
So now cACTION_CONTROL_CLIENT_START_MONITORING_REQUEST  always sent with  ip=0

morantr committed:
controller: Fix IP notification destination address

The address which we have sent the IP update notification has been the VAP
mac address instead of the radio mac.
Also, The agent that we send the notification to is the Agent that sends us the
DHCP event, which is not correct since the DHCP server runs on the gw and the
client might be connected to another IRE.

Change the destination radio and the destination Agent to the correct ones.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Fix unset ipv4 field

Fix t he ipv4 field on the start monitoring message that has not been initialized.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Add zero IP option for switching IP to 255.255.255.255

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3026,46 +3026,53 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
-                LOG(ERROR) << "set node ipv4 failed";
-            }
+        if (database.get_node_type(client_mac) != beerocks::TYPE_CLIENT) {
+            LOG(DEBUG) << "node " << client_mac << " is not of TYPE_CLIENT";
```
```suggestion
            LOG(INFO) << "Ignoring DHCP notification for mac " << client_mac << ", as it's not a client";
```

morantr replied:
done.

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3026,46 +3026,53 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
-                LOG(ERROR) << "set node name failed";
-            }
+        if (!database.set_node_ipv4(client_mac, ipv4)) {
+            LOG(ERROR) << "set node ipv4 failed";
```
Does it makes sense to continue in this case?
If you don't have an IP, what are you actually sending to the agent?

morantr replied:
Actually it cannot be failed if the previous db setter success.
The set ipv4 failed if the client has not been found on the db.
I'll remove the failure handling.

vitalybu reviewed (approved):
Few more comments, but approving as it should also work as is
